src/main/java/ch/njol/skript/expressions/ExprBookPages.java

```diff
@@ -19,99 +19,189 @@
+
+	private Expression<ItemType> items;
 	@Nullable
 	private Expression<Number> page;
-	
-	@SuppressWarnings("null")
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		if (matchedPattern == 0 || matchedPattern == 1) {
+			items = (Expression<ItemType>) exprs[0];
+		} else {
+			if (matchedPattern == 2) {
+				page = (Expression<Number>) exprs[0];
+				items = (Expression<ItemType>) exprs[1];
+			} else {
+				items = (Expression<ItemType>) exprs[0];
+				page = (Expression<Number>) exprs[1];
+			}
+		}
+		return true;
+	}
+
+	@Override
 	@Nullable
+	@SuppressWarnings("null")
+	protected String[] get(Event event) {
+		List<String> pages = new ArrayList<>();
+		for (ItemType itemType : items.getArray(event)) {
+			if (itemType == null || !(itemType.getItemMeta() instanceof BookMeta))
+				continue;
+			BookMeta bookMeta = (BookMeta) itemType.getItemMeta();
+			if (isAllPages()) {
+				pages.addAll(bookMeta.getPages());
+			} else {
+				Number pageNumber = page.getSingle(event);
+				if (pageNumber == null)
+					continue;
+				int page = pageNumber.intValue();
+				if (page <= 0 || page > bookMeta.getPageCount())
+					continue;
+				pages.add(bookMeta.getPage(page));
+			}
+		}
+		return pages.toArray(new String[0]);
+	}
+
 	@Override
-	protected String[] get(Event e) {
-		ItemStack itemStack = book.getSingle(e).getRandom();
-		if (itemStack == null || !bookItem.isOfType(itemStack))
-			return null;
-		List<String> pages = ((BookMeta) itemStack.getItemMeta()).getPages();
-		if (page != null){
-			Number pageNumber = page.getSingle(e);
-			if (pageNumber == null){
+	@Nullable
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		switch (mode) {
+			case RESET:
+			case DELETE:
+				return CollectionUtils.array();
+			case SET:
+				return CollectionUtils.array(isAllPages() ? String[].class : String.class);
+			case ADD:
+				return isAllPages() ? CollectionUtils.array(String[].class) : null;
+			default:
 				return null;
+		}
+	}
+
+	@Override
+	@SuppressWarnings("ConstantConditions")
+	public void change(Event event, Object @Nullable [] delta, ChangeMode mode) {
+		if ((mode == ChangeMode.SET || mode == ChangeMode.ADD) && delta == null)
+			return;
+		ItemType[] itemTypes = items.getArray(event);
+		int page = !isAllPages() ? this.page.getOptionalSingle(event).orElse(-1).intValue() : -1;
+		String[] newPages = delta == null ? null : new String[delta.length];
+
+		if (newPages != null)
+			for (int i = 0; i < delta.length; i++)
+				newPages[i] = delta[i] + "";
+
+		for (ItemType itemType : itemTypes) {
+			if (itemType == null || !(itemType.getItemMeta() instanceof BookMeta))
+				continue;
+
+			BookMeta bookMeta = (BookMeta) itemType.getItemMeta();
+			List<String> pages = null;
+			if (isAllPages()) {
+				switch (mode) {
+					case DELETE:
+					case RESET:
+						pages = Collections.singletonList("");
```

Comment:
Is it allowed to just be an empty list

Reply:
https://github.com/SkriptLang/Skript/pull/5025#discussion_r943103754